keps/369-job-interface/README.md

```diff
@@ -0,0 +1,329 @@
+implementing the interface people can also build their controllers.
+- This will increase the code complexity by wrapping the original Jobs.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Part I: Job Interface
+
+We will define a new interface named Job, this should be implemented by custom job-like applications:
+
+```golang
+type Job interface {
+  // Suspend instructs whether the job is suspended or not.
+  Suspend() bool
+  // Start means starting the job, like unsuspending the job.
+  Start(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // Stop means stopping the job, like suspending the job.
+  Stop(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
+  InjectNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // RestoreNodeAffinity will restore the original node affinity of job.
+  RestoreNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // Finished instructs whether the job is completed/failed or not.
+  Finished() bool
+  // Finish will update the workload to finished.
+  Finish(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // EquivalentToWorkload validates whether the workload is semantically equal to the job.
+  EquivalentToWorkload(wl kueue.Workload) bool
+  // CreateWorkload will create a workload constructing by the job.
+  CreateWorkload(ctx context.Context, client client.Client, scheme *runtime.Scheme) (*kueue.Workload, error)
+  // Priority returns the job's priority
+  Priority(ctx context.Context, client client.Client) (string, int32, error)
+  // QueueName returns the queue name the job enqueued.
+  QueueName() string
+  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
+  Ignored() bool
+  // PodsCount returns the pods number job requires.
+  PodsCount() int32
+}
+```
+
+Also defined another interface for all-or-nothing scheduling required job-like applications:
+
+```golang
+type GangSchedulingJob interface {
+  // GangSchedulingSucceeded instructs whether job is scheduled successfully in gang-scheduling.
+  GangSchedulingSucceeded() bool
```

Comment:
`GangSchedulingJob` doesn't sound appropriate to me. Since we don't use the expression `GangSchedulling` in the document and kueue doesn't guarantee gang-schedule all Pods to Nodes.

https://github.com/kubernetes-sigs/kueue/blob/160021fcd342f0b010e4d0d5e64ea634dba54df2/docs/tasks/setup_sequential_admission.md

However, I'm not sure what name is appropriate.

Reply:
waitForPodsReady is currently a Kueue level configuration, so it doesn't make sense to have it as part of the GenericJobInterface for now.

Reply:
> waitForPodsReady is currently a Kueue level configuration, so it doesn't make sense to have it as part of the GenericJobInterface for now.

I agree. I wanted to ask whether `Gang` is a general term or not.

Reply:
My original thought is this may also need by other jobs, by implementing like https://github.com/kubernetes-sigs/kueue/blob/3268949a656a18b779c6d5369de8a1a92034cef4/pkg/controller/workload/job/job_controller.go#L255-L258

But we can extend the interface if needed, removed then.

Reply:
ah, I see what you mean now. Let's better be more explicit and make it `PodsReady() bool`

Reply:
I see too. @alculquicondor's idea sounds good to me.

Reply:
Add a new interface SequenceJob to containing `PodsReady()`

Reply:
Does it need to be a separate interface?

Reply:
Not every job needs this, I treat it as an advanced feature, so maybe it's good to keep the interface as small as possible.

Reply:
What would we do with a job that doesn't implement the interface if the kueue is configured with `waitForReadyPods: true`?